src/hotspot/share/opto/intrinsicnode.cpp

```diff
@@ -112,3 +114,102 @@ SignumFNode* SignumFNode::make(PhaseGVN& gvn, Node* in) {
+      // compress(x, 1 << n) == (x >> n & 1)
+      if (phase->type(mask->in(1))->higher_equal(TypeLong::ONE)) {
+        Node* rshift = phase->transform(new RShiftLNode(in(1), mask->in(2)));
+        return new AndLNode(rshift, phase->makecon(TypeLong::ONE));
+      // compress(x, -1 << n) == x >>> n
+      } else if (phase->type(mask->in(1))->higher_equal(TypeLong::MINUS_1)) {
+        return new URShiftLNode(in(1), mask->in(2));
+      }
+    }
+    // compress(expand(x, m), m) == x & compress(m, m)
+    if (src->Opcode() == Op_ExpandBits &&
+        src->in(2) == mask) {
+      Node* compr = phase->transform(new CompressBitsNode(mask, mask, TypeLong::LONG));
+      return new AndLNode(compr, src->in(1));
+    }
+  }
+  return NULL;
+}
+
+Node* compress_expand_identity(PhaseGVN* phase, Node* n) {
+  BasicType bt = n->bottom_type()->array_element_basic_type();
+  // compress(x, 0) == 0
+  if(phase->type(n->in(2))->higher_equal(TypeInteger::zero(bt))) return n->in(2);
+  // compress(x, -1) == x
+  if(phase->type(n->in(2))->higher_equal(TypeInteger::minus_1(bt))) return n->in(1);
+  return n;
+}
+
+Node* CompressBitsNode::Identity(PhaseGVN* phase) {
+  return compress_expand_identity(phase, this);
+}
+
+
+Node* ExpandBitsNode::Ideal(PhaseGVN* phase, bool can_reshape) {
+  Node* src = in(1);
+  Node* mask = in(2);
+  if (bottom_type()->isa_int()) {
+    if (mask->Opcode() == Op_LShiftI && phase->type(mask->in(1))->is_int()->is_con()) {
+      // expand(x, 1 << n) == (x & 1) << n
+      if (phase->type(mask->in(1))->higher_equal(TypeInt::ONE)) {
+        Node* andnode = phase->transform(new AndINode(in(1), phase->makecon(TypeInt::ONE)));
+        return new LShiftINode(andnode, mask->in(2));
+      // expand(x, -1 << n) == x << n
+      } else if (phase->type(mask->in(1))->higher_equal(TypeInt::MINUS_1)) {
+        return new LShiftINode(in(1), mask->in(2));
+      }
+    }
+    // expand(compress(x, m), m) == x & m
+    if (src->Opcode() == Op_CompressBits &&
+        src->in(2) == mask) {
+      return new AndINode(src->in(1), mask);
+    }
+  } else {
+    assert(bottom_type()->isa_long(), "");
+    if (mask->Opcode() == Op_LShiftL && phase->type(mask->in(1))->is_long()->is_con()) {
+      // expand(x, 1 << n) == (x & 1) << n
+      if (phase->type(mask->in(1))->higher_equal(TypeLong::ONE)) {
+        Node* andnode = phase->transform(new AndLNode(in(1), phase->makecon(TypeLong::ONE)));
+        return new LShiftLNode(andnode, mask->in(2));
+      // expand(x, -1 << n) == x << n
+      } else if (phase->type(mask->in(1))->higher_equal(TypeLong::MINUS_1)) {
+        return new LShiftLNode(in(1), mask->in(2));
+      }
+    }
+    // expand(compress(x, m), m) == x & m
+    if (src->Opcode() == Op_CompressBits &&
+        src->in(2) == mask) {
+      return new AndLNode(src->in(1), mask);
+    }
+  }
+  return NULL;
+}
+
+Node* ExpandBitsNode::Identity(PhaseGVN* phase) {
```

Comment:
I also suggest adding a boolean if `compress_expand_identity` if you add rules which don't apply to both equally.

Here is possible type-propagation logic for compress and expand:

```
let SIGN_BIT = (((IntOrLong)-1)>>>1)+1  (bit 31 or 63)
let MAX_POS  = (((IntOrLong)-1)>>>1)
lot BITS = 1+bitCount(MAX_POS)  (32 or 64)
if (both x, m are con) {
  // maybe use these rules, by porting the Java code to C++
  compress(CON[x], CON[m]) ] = CON[portable_compress(x,m)]
  expand(CON[x], CON[m]) ] = CON[portable_expand(x,m)]
  // see also https://stackoverflow.com/questions/38938911/portable-efficient-alternative-to-pdep-without-using-bmi2
} else if (m is CON[m] && m != -1) {
  //compress(x, -1) = x   //identity handled elsewhere
  //expand(x, -1) = x     //identity handled elsewhere
  let bitc = bitCount(m)
  LO[ compress(x, CON[m]) ] = 0  //sign bit is never set
  HI[ compress(x, CON[m]) ] = ((1L<<bitc)-1)  //unsigned bit-support
  LO[ expand(x, CON[m]) ] = (m >= 0) ? 0 : SIGN_BIT //sign bit might be set alone
  HI[ expand(x, CON[m]) ] = (m >= 0) ? m : m ^ SIGN_BIT
  // could improve a little by looking TYPE[x], but do not bother
} else {
  // estimate maximum possible weight of m (in 0..63)
  let  maxbitc = BITS   if (LO[m] < 0 && HI[m] >= -1)  // could be -1
  else maxbitc = BITS-1 if (LO[m] < 0 || HI[m] == MAX_POS)  // <0 or maxint
  else maxbitc = BITS-1 - numberOfLeadingZeros(HI[m])
  LO[ compress(x, m) ] = (maxbitc == 64 && LO[x] < 0) ? SIGN_BIT : 0
  HI[ compress(x, m) ] = (maxbitc >= 63) ? HI[x] : MIN(HI[x], (1L<<maxbitc)-1)
  LO[ expand(x, m) ] = (LO[m] >= 0) ? 0 : SIGN_BIT
  HI[ expand(x, m) ] = (LO[m] >= 0) ? HI[m] : MAX_POS
}
```

The operands of compress and expand are inherently unsigned bitmasks, so the signed type system of C2 gets in the way.  In the future, a somewhat more thorough job could be done if we had bitwise types as well in C2.  For that that would mean, see https://bugs.openjdk.java.net/browse/JDK-8001436

Reply:
I have handled these transformation separately in ideal/identity and value routines.

Reply:
Hi @rose00 , I have extended the value routines with value propagation logic.